### PR TITLE
openclaw/app: sanitize MCP image tool messages

### PR DESCRIPTION
**Pull request by WineChord**
## Summary
- strip inline image data from MCP image tool messages before replaying them to the model
- keep text metadata in the tool message and attach supported images through content parts when budget allows
- return a sanitized tool message even when the attachment budget is exhausted or image decoding fails

## Validation
- go test ./app -run 'TestMCPImageResultMessages|TestExtractMCPImages|TestUnwrapMCPResultContent|TestMCPImageFormatFromMime'\n- go test ./app

### CODE REVIEW

File: openclaw/app/mcp_images.go [hunk 1]
```diff
@@ -83,6 +89,85 @@ func mcpImageResultMessages(
...
+		out = append(out, item)
+	}
+	return out
+}
+
+func sanitizedMCPResultJSON(
+	result any,
+	items []mcpContentItem,
+) (string, bool) {
+	body, err := json.Marshal(result)
```
Comment (Flash-LHR) [inline openclaw/app/mcp_images.go]: `json.Marshal` HTML escapes text in sanitized MCP image results, changing tool output seen by the model. Serialize this content with `SetEscapeHTML(false)`.
<details><summary>中文</summary>
`json.Marshal` 会对脱敏后的 MCP 图片结果文本做 HTML 转义，改变模型看到的工具输出。请用 `SetEscapeHTML(false)` 序列化这段内容。
</details>

Comment (WineChord) [reply openclaw/app/mcp_images.go]: fixed

File: openclaw/app/mcp_images.go [hunk 2]
```diff
@@ -83,6 +90,116 @@ func mcpImageResultMessages(
 	return []model.Message{defaultMsg, userMsg}, nil
 }
 
+func mcpImageUserContent(result any) string {
+	path := mcpImageSavedPath(result)
+	if path == "" {
+		return mcpImagesUserContent
+	}
+	return mcpImagesUserContent + " Saved file: " + path +
```
Comment (Rememorio) [inline openclaw/app/mcp_images.go]: **P1: Do not promote result-controlled paths into user instructions**

Previously this synthetic user message was fixed text. It now copies `screenshotPath` from any result containing an MCP image and instructs the model, as `RoleUser`, to invoke a local file-reading tool. Because this callback is registered globally, an external MCP tool can supply that field and steer the model toward a path permitted by `image_inspect`, crossing the tool-output trust boundary.

Only emit this instruction when the path has trusted provenance from the native browser screenshot operation and was validated for the same call; otherwise leave it as tool-role metadata or omit it. A fake external MCP result containing an arbitrary `screenshotPath` should not produce a user-role file-inspection instruction.

<details><summary>中文</summary>

**不要将结果可控路径提升为用户指令**

此前这条合成的 user 消息是固定文本；现在代码会从任何包含 MCP 图片的结果中读取 `screenshotPath`，并以 `RoleUser` 身份指示模型调用本地文件读取工具。由于该回调是全局注册的，外部 MCP 工具可以伪造这个字段，引导模型读取 `image_inspect` 权限范围内的路径，跨越了工具输出的信任边界。

只有当路径确实来自受信任的原生浏览器截图操作，并且已针对本次调用完成校验时，才应生成这条指令；其他情况应保留为 tool 角色元数据或直接省略。建议增加伪造外部 MCP 结果的测试，确保任意 `screenshotPath` 不会生成 user 角色的文件检查指令。

</details>

File: openclaw/app/mcp_images.go [hunk 3]
```diff
@@ -83,6 +90,116 @@ func mcpImageResultMessages(
...
+		return nil
+	}
+	out := make([]mcpContentItem, 0, len(items))
+	for _, item := range items {
+		if strings.ToLower(strings.TrimSpace(item.Type)) !=
+			mcpContentTypeImage {
+			out = append(out, item)
+			continue
+		}
+		if _, ok := mcpImageFormatFromMime(item.MimeType); !ok {
```
Comment (Rememorio) [inline openclaw/app/mcp_images.go]: **P1: Redact unsupported image formats too**

Unsupported image MIME types are appended unchanged, so their inline data remains in the tool message. If the result contains only such images, `countMCPImageCandidates` also returns zero and the callback falls back to the completely unsanitized default message. Large TIFF, SVG, or vendor-specific image payloads can therefore still exhaust the model context.

Redact `data` for every item whose type is `image`, while keeping the supported-format check only for attachment decoding. Cover both unsupported-only and mixed supported/unsupported results in tests.

<details><summary>中文</summary>

**不支持的图片格式也应脱敏**

当前逻辑会原样保留 MIME 类型不受支持的图片，因此其内联数据仍会进入工具消息。如果结果只包含这类图片，`countMCPImageCandidates` 还会返回零，回调最终使用完全未脱敏的默认消息。较大的 TIFF、SVG 或厂商自定义图片仍可能耗尽模型上下文。

建议对所有 `type` 为 `image` 的条目统一移除 `data`，仅在生成附件时检查格式是否受支持。测试应同时覆盖“只有不支持格式”和“支持与不支持格式混合”两种结果。

</details>

File: openclaw/app/mcp_images.go [hunk 4]
```diff
@@ -83,6 +90,116 @@ func mcpImageResultMessages(
...
+
+	var envelope map[string]json.RawMessage
+	if err := json.Unmarshal(body, &envelope); err != nil {
+		return "", false
+	}
+	if _, ok := envelope["content"]; !ok {
+		return "", false
+	}
+
+	itemBytes, err := marshalModelVisibleJSON(items)
```
Comment (Rememorio) [inline openclaw/app/mcp_images.go]: **P1: Preserve non-image MCP content fields**

Before this change, the default tool message preserved the complete serialized result. The new sanitizer replaces `content` with `items`, but `mcpContentItem` only retains `type`, `data`, and `mimeType`; standard MCP text uses `text`, and resource content has additional fields. Any mixed result containing a supported image therefore loses its text or resource payload during replay.

Please sanitize the original content as `json.RawMessage` or maps and replace only image `data` fields. Add a regression case containing `{"type":"text","text":"visible"}` alongside an image and assert that `text` survives unchanged.

<details><summary>中文</summary>

**保留非图片 MCP 内容字段**

修改前，默认工具消息会完整保留结果的序列化内容。现在脱敏逻辑用 `items` 替换整个 `content`，但 `mcpContentItem` 只保留 `type`、`data` 和 `mimeType`；标准 MCP 文本使用的是 `text` 字段，resource 类型还包含更多字段。因此，只要混合结果中存在受支持的图片，重放时文本或资源内容就会被丢弃。

建议基于原始 `json.RawMessage` 或 map 做局部修改，只替换图片项的 `data`。同时增加一个包含 `{"type":"text","text":"visible"}` 和图片的回归用例，确认 `text` 能原样保留。

</details>

File: openclaw/app/mcp_images.go [hunk 5]
```diff
@@ -32,6 +33,7 @@ const (
 	mcpImageDetailAuto = "auto"
 
 	mcpImagesUserContent = "MCP tool returned image(s)."
+	mcpImageDataOmitted  = "[omitted: image data attached separately]"
```
Comment (Rememorio) [inline openclaw/app/mcp_images.go]: **P2: Do not claim an omitted image was attached**

This marker says the image was attached separately, but the same sanitized message is returned when the attachment budget is exhausted or every image fails base64 decoding. In those branches no user message or image content part is added, leaving the model with a false indication that the image is available.

Use a neutral marker such as `[omitted: image data]`, or select the wording only after extraction determines whether an attachment was actually added. Assert the marker semantics in both the zero-budget and decode-failure tests.

<details><summary>中文</summary>

**没有附件时不要声称图片已附加**

该占位文本表示图片已经单独附加，但附件预算耗尽或所有图片都无法完成 base64 解码时，代码也会返回同样的脱敏消息。这些分支并不会添加 user 消息或图片内容块，模型却会误以为图片仍然可用。

建议改用 `[omitted: image data]` 之类的中性文本，或者在完成解码并确认实际添加附件后再选择对应文案。零预算和解码失败测试都应校验占位文本与真实附件状态一致。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 4 inline comments below. The comments focus on issues worth addressing before merge.

Reviewed the MCP image replay sanitizer and browser screenshot-path propagation. Four actionable risks remain: mixed MCP content loses standard fields, unsupported image formats bypass redaction, result-controlled paths are promoted into user-role instructions, and omitted images are described as attached when no attachment exists. Validation should cover standard MCP text-plus-image payloads, unsupported MIME types, untrusted screenshotPath values, exhausted budgets, and decode failures.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 4 条行内评论。评论聚焦在合并前值得处理的问题。

已审查 MCP 图片结果的脱敏与重放逻辑，以及浏览器截图路径的传递方式。目前有四项需要处理的风险：混合 MCP 内容中的标准字段会丢失、不支持的图片格式会绕过脱敏、工具结果可控的路径被提升为 user 角色指令，以及没有实际附件时仍声称图片已单独附加。建议补充标准 text+image 载荷、不支持的 MIME、恶意 screenshotPath、附件预算耗尽和解码失败等场景的验证。

</details>

File: openclaw/app/mcp_images.go (hunk 2 above)
Comment (WineChord) [reply openclaw/app/mcp_images.go]: fixed

File: openclaw/app/mcp_images.go (hunk 3 above)
Comment (WineChord) [reply openclaw/app/mcp_images.go]: fixed

File: openclaw/app/mcp_images.go (hunk 4 above)
Comment (WineChord) [reply openclaw/app/mcp_images.go]: fixed

File: openclaw/app/mcp_images.go (hunk 5 above)
Comment (WineChord) [reply openclaw/app/mcp_images.go]: fixed

File: openclaw/app/mcp_images.go (hunk 2 above)
Comment (Rememorio) [reply openclaw/app/mcp_images.go]: Confirmed fixed: the user-role message is now path-independent, and the forged `screenshotPath` regression test passes.

<details><summary>中文</summary>

确认已修复：用户角色消息现已与路径无关，伪造 `screenshotPath` 的回归测试也已通过。

</details>

File: openclaw/app/mcp_images.go (hunk 3 above)
Comment (Rememorio) [reply openclaw/app/mcp_images.go]: Confirmed fixed. All image items are now sanitized regardless of MIME type, and the unsupported-only and mixed-format cases are covered by tests.

<details><summary>中文</summary>

已确认修复。现在所有图片条目都会脱敏，不受 MIME 类型限制，并且测试已覆盖“仅不支持格式”和“混合格式”场景。

</details>

File: openclaw/app/mcp_images.go (hunk 4 above)
Comment (Rememorio) [reply openclaw/app/mcp_images.go]: Confirmed fixed. The sanitizer now preserves non-image fields while replacing only image data, and the mixed text/resource/image regression test passes.

<details><summary>中文</summary>

确认已修复。脱敏逻辑现在会保留非图片字段，仅替换图片数据，且文本、资源与图片混合场景的回归测试已通过。

</details>

File: openclaw/app/mcp_images.go (hunk 5 above)
Comment (Rememorio) [reply openclaw/app/mcp_images.go]: Fixed. The marker is now neutral, and the zero-budget and decode-failure paths are both covered by assertions.

<details><summary>中文</summary>

已修复。占位文本现在是中性的，零预算和解码失败路径也都有相应断言覆盖。

</details>

File: openclaw/internal/browser/tool.go [hunk 6]
```diff
@@ -1570,6 +1570,7 @@ func (t *Tool) handleScreenshot(
 		t.evaluateEnabled,
 	)
 	result.TargetID = strings.TrimSpace(in.TargetID)
+	result.ScreenshotPath = filename
```
Comment (Rememorio) [inline openclaw/internal/browser/tool.go]: **P2: Do not report a screenshot path after a failed capture**

`ScreenshotPath` is assigned even when `compactBrowserErrorResult` has recognized a browser crash delivered as an ordinary content payload. With `screenshot_dir` configured, `resolveScreenshotFilename` generates a non-empty filename before the call, so this path is returned even though the screenshot was never written.

Previously the result made no artifact claim; now models and downstream consumers can attempt to inspect a nonexistent file. Populate this field only after confirming the capture succeeded, and add a crash-with-`screenshot_dir` test asserting that `ScreenshotPath` stays empty.

<details><summary>中文</summary>

**截图失败时不应返回截图路径**

即使 `compactBrowserErrorResult` 已识别出以普通内容形式返回的浏览器崩溃，这里仍会设置 `ScreenshotPath`。配置了 `screenshot_dir` 时，`resolveScreenshotFilename` 会在调用驱动前生成非空文件名，因此截图实际未写入也会返回该路径。

旧结果不会宣称已生成截图制品；现在模型或下游消费者可能继续读取一个不存在的文件。建议仅在确认截图成功后设置该字段，并增加“配置 `screenshot_dir` 且浏览器崩溃”的测试，断言 `ScreenshotPath` 为空。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 1 inline comment below. The comments focus on issues worth addressing before merge.

Reviewed MCP image sanitization, attachment-budget fallbacks, browser result serialization, and related tests. The sanitization paths preserve non-image metadata and omit inline image data as intended, but the new browser screenshot path can falsely advertise an artifact after a browser crash; add a failure-path assertion before relying on this field.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 1 条行内评论。评论聚焦在合并前值得处理的问题。

已检查 MCP 图片净化、附件预算降级、浏览器结果序列化及相关测试。图片内联数据会按预期移除，非图片元数据也能保留；但新增的截图路径可能在浏览器崩溃后错误地宣称制品已生成，需要补充失败路径校验。

</details>

File: openclaw/internal/browser/tool.go (hunk 6 above)
Comment (WineChord) [reply openclaw/internal/browser/tool.go]: fixed in d1bebfe09: detect crash payloads before normalization and only report ScreenshotPath after a successful capture; crash-with-screenshot_dir regression added

**Review by Flash-LHR — APPROVED** [review]
LGTM
